src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5313,6 +5312,58 @@
             var inferredResult = InferResultNullability(operatorKind, method, returnType, leftType, rightType);
 
             return inferredResult;
+        }
+
+        MethodSymbol ReInferBinaryOperator(
+            SyntaxNode syntax,
+            MethodSymbol method,
+            BoundExpression leftOperand,
+            BoundExpression rightOperand,
+            TypeWithState leftUnderlyingType,
+            TypeWithState rightUnderlyingType)
+        {
+            TypeSymbol methodContainer = method.ContainingType;
+            MethodSymbol reinferredMethod;
+
+            if (!method.GetIsNewExtensionMember())
+            {
+                TypeSymbol asMemberOfType = getTypeIfContainingType(methodContainer, leftUnderlyingType.Type, leftOperand) ??
+                    getTypeIfContainingType(methodContainer, rightUnderlyingType.Type, rightOperand) ?? methodContainer;
+                reinferredMethod = (MethodSymbol)AsMemberOfType(asMemberOfType, method);
+            }
+            else if (method.ContainingType.Arity != 0)
+            {
+                NamedTypeSymbol extension = method.OriginalDefinition.ContainingType;
+                var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+                var inferenceResult = MethodTypeInferrer.Infer(
+                    _binder,
+                    _conversions,
+                    extension.TypeParameters,
+                    extension,
+                    method.OriginalDefinition.ParameterTypesWithAnnotations,
+                    method.OriginalDefinition.ParameterRefKinds,
+                    [new BoundExpressionWithNullability(leftOperand.Syntax, leftOperand, leftUnderlyingType.ToTypeWithAnnotations(compilation).NullableAnnotation, leftUnderlyingType.Type),
```

Comment:
>BoundExpressionWithNullability [](http://example.com/codeflow?start=25&length=30)

In analysis of invocations (`VisitCall`/`VisitArguments`), we use `GetArgumentsForMethodTypeInference` to get inputs to `MethodTypeInferrer.Infer`. Do we need the same thing here (it has extra cases to deal with lambda, collection expressions and typeless expressions)?

Reply:
```GetArgumentsForMethodTypeInference``` operates in terms of ```VisitResult``` for arguments, this function gets ```TypeWithState``` instead. Moreover, the passed in ```TypeWithState``` is getting adjusted by the caller for lifted scenarios. Given that, it doesn't look trivial to switch from ```TypeWithState``` to ```VisitResult``` as an input. I think the current implementation is going to work for majority of scenarios, I will leave a prototype comment to follow up.